### PR TITLE
[WIP] LLVM Basic JIT

### PR DESCRIPTION
**Pull request by bollu**
Working on the JIT by adapting code from Brainfuck

TODO
- [ ] JIT booleans
- [ ] JIT boolean operations

- [x] JIT basic integers
- [ ] JIT integer operations

- [ ] JIT pure functions
- [ ] JIT closures
- [ ] Figure out GC (semi-herculean)
- [ ] have JIT re-enter into Rust code for network, file system

### CODE REVIEW

**Comment by bollu** [comment]
Currently generates IR:

```cpp
@BalloonInt = constant i32 0
@BalloonInt.1 = constant i32 2
@BalloonInt.2 = constant i32 4

; Function Attrs: argmemonly nounwind
declare void @llvm.memset.p0i8.i32(i8* nocapture writeonly, i8, i32, i32, i1) #0

declare i8* @malloc(i32)

declare void @free(i8*)

declare i32 @write(i32, i8*, i32)

declare i32 @putchar(i32)

declare i32 @getchar()

define { i32, i64 } @entry_point() {
init:
}

define { i32, i64 } @box_i64(i64 %input) {
box_int_fn_entry:
  %box = alloca { i32, i64 }
  %box_slot = getelementptr inbounds { i32, i64 }, { i32, i64 }* %box, i32 0, i32 1
  store i64 %input, i64* %box_slot
  %int_tag = load i32, i32* @BalloonInt
  %box_tag = getelementptr inbounds { i32, i64 }, { i32, i64 }* %box, i32 0, i32 0
  store i32 %int_tag, i32* %box_tag
  ret { i32, i64 }* %box
}

attributes #0 = { argmemonly nounwind }
```

I'll generalise the `box` to box `f64` and `bool`. Will also build an `unbox` and `add`. Then I can actually JIT some addition `:)`. Hopefully by day after.

File: .travis.yml [hunk 1]
```diff
@@ -8,8 +8,8 @@ rust:
   - nightly
 
 before_install:
-  - deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-4.0 main
-  - deb-src http://apt.llvm.org/trusty/ llvm-toolchain-trusty-4.0 main
+  - sudo echo "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-4.0 main" >> /etc/apt/sources.list
```
Comment (polybuildr) [inline .travis.yml]: The redirect loses the sudo. So you need something like:

```sh
echo "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-4.0 main" | sudo tee -a /etc/apt/sources.list
``

**Comment by bollu** [comment]
@polybuildr review, please

File: src/llvm_interpreter.rs [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+            let leftbox = compile_expr(module, bb, box_unbox_functions, &leftexpr.data);
+            let rightbox = compile_expr(module, bb, box_unbox_functions, &rightexpr.data);
+
+            let leftval = LLVMBuildCall(builder.builder, box_unbox_functions.unbox_i64,
+                                        [leftbox].as_mut_ptr(), 1, module.new_string_ptr("leftval"));
+            let rightval = LLVMBuildCall(builder.builder, box_unbox_functions.unbox_i64,
+                                         [rightbox].as_mut_ptr(), 1, module.new_string_ptr("rightval"));
+
+            let finalval = match op {
+                Add => LLVMBuildNSWAdd(builder.builder, leftval, rightval, module.new_string_ptr("addval")),
```
Comment (polybuildr) [inline src/llvm_interpreter.rs]: This isn't matching against `BinaryOp::Add` and instead seems to be wildcard matching creating a new binding `Add`, which means none of the other patterns ever get matched.

**Review by polybuildr — CHANGES_REQUESTED** [review]
Initial comments:

## Installation
So I installed `llvm-4.0` and `llvm-4.0-tools` to run this. I think that installs all the other packages, or some might not be required - but things run fine with just those two.

I found it rather odd that llvm doesn't give you their apt key from an https source though. :/

## Compiling

There seem to be a rather large number of warnings on compilation: unused imports, dead code, non-snake cased identifiers, unused variables, one unreachable pattern and one unnecessary `mut`. Since this is prototype code, it doesn't matter much, but please eventually clean up. :P

## Running
So, to confirm, I run `cargo run -- --repl-llvm` and type in `10 + 20` and hit enter.
This gives lots of generated llvm IR, I think followed by this:

```
@@@@@@ ERROR: 
@@@@@ Creating Execution Engine...
@@@@@ Execution Engine Created.
@@@@@ LLVMRunFunction succeesed
@@@@ Result of Function call:
result ptr : |0x7fc4da22a8a0|
*(result ptr) : |LLVMBoxedStructRepr { tag: 30, raw: 50 }|
```

The 50 seems to be the result of the addition, which is awesome.

Trying anything apart from `+` still gives addition though. Is this by design temporarily? If so, should probably rename the commit to talk only about addition. However, I think it's a pattern matching issue, so I'm leaving a comment inline about that.

**Review by polybuildr — APPROVED** [review]

**Comment by polybuildr** [comment]
I don't understand much LLVM related stuff, but things look okay overall to me now. Merging. Thanks for the contribution! :D

**Pull request merged**